include/boost/thread/pthread/condition_variable.hpp

```diff
@@ -224,18 +245,22 @@ namespace boost
         template<typename lock_type,typename duration_type>
         bool timed_wait(lock_type& m,duration_type const& wait_duration)
         {
-            return timed_wait(m,get_system_time()+wait_duration);
+            if (wait_duration.is_pos_infinity())
+            {
+                wait(m); // or do_wait(m,detail::timeout::sentinel());
+                return true;
+            }
+            if (wait_duration.is_special())
+            {
+                return true;
+            }
+            return do_wait_until(m,detail::timespec_plus_internal_clock(wait_duration));
         }
 
         template<typename lock_type,typename predicate_type>
         bool timed_wait(lock_type& m,boost::system_time const& abs_time, predicate_type pred)
         {
-            while (!pred())
-            {
-                if(!timed_wait(m, abs_time))
-                    return pred();
-            }
-            return true;
+            return do_wait_until(m,detail::timespec_to_internal_clock(abs_time),pred);
```

Comment:
system_time and internal_clock can not be converted when the clock is steady

Reply:
I believe the timespec_to_internal_clock function handles this case. When the monotonic clock is enabled, it takes the difference between the abs_time and the current system time, and then adds that to the current steady clock, and that's what it returns. That ensures that we wait for the correct amount of time, even if the system clock jumps while we're waiting.